watchdir.1
.TH watchdir 1 "September 2021" Linux "User Manuals"
.SH NAME
watchdir \- watch for changes in a directory tree
.SH SYNOPSIS
.B watchdir
[\fIOPTION\fR]
\fIDIRECTORY\fR
\fIOUTPUTDIR\fR
.br
.B watchdir
[\fI\-h\fR|\fI\-V\fR]

.SH DESCRIPTION
.B watchdir
monitors the given
.I DIRECTORY
and all its subdirectories for changes.  At regular intervals, the names of
the files and directories which have changed are written to a uniquely named
file in the
.I OUTPUTDIR
directory:

.IP * 3
If a file is changed or created, its filename will be listed.
.IP * 3
If a file is removed, this is treated as a change to the directory it was
in, so the directory will be listed.
.IP * 3
If a file is renamed, this is treated as a removal and creation, so both the
file's new path and the directory containing its old name will be listed.

.PP

Changes to directories are treated in the same way:

.IP * 3
If a directory is created, the new directory will be listed.
.IP * 3
If a directory is removed, its parent directory will be listed.
.IP * 3
If a directory is renamed, its old parent directory will be listed, and its
new path will be listed.

.PP

All paths are given relative to the top-level
.IR DIRECTORY .
Directory names always end in /, which means that just
.B /
on its own in a change file means that the top-level directory itself
changed (i.e. a file or directory was removed from it).

.PP

The change files in
.I OUTPUTDIR
are given names of the form
.B YYYYMMDD-hhmmss.pid
where
.I YYYY
is the 4-digit year,
.I MM
is the 2-digit month,
.I DD
is the 2-digit day of the month,
.IR hh ", " mm ", " ss
are 2-digit hours, minutes, and seconds in the 24-hour clock, and
.I pid
is the process ID of
.BR watchdir .

.PP

Note that
.I OUTPUTDIR
must not be a subdirectory of
.IR DIRECTORY .


.SH OPTIONS

.TP
.BR \-i ", " "\-\-dump\-interval SEC"
Write change files every
.I SEC
seconds.  The default is 30 seconds.  No change files are written when there
are no changes to report.
.TP
.BR \-f ", " "\-\-full\-scan\-interval SEC"
Rescan
.I DIRECTORY
every
.I SEC
seconds, to pick up any changes that might have been missed.  The default
is 7200 seconds (2 hours).
.TP
.BR \-e ", " "\-\-exclude PATTERN"
A
.BR glob (7)
pattern to exclude.  Any file or directory, in any subdirectory of
.IR DIRECTORY ,
which matches this pattern will be ignored.

This option can be specified multiple times.

The default is to exclude
.BR "*.tmp" " and " "*~" .
.TP
.BR \-r ", " "\-\-recursion\-depth NUM"
Descend no more than
.I NUM
directories deep into
.IR DIRECTORY .
The default is 20.  It is not advisable to set this too high as it may cause
excessive consumption of system resources.
.TP
.BR \-q ", " "\-\-queue\-run\-interval SEC"
Process the
.BR inotify (7)
change queue every
.I SEC
seconds.  The default is 2 seconds.  This will rarely need to be changed.
.TP
.BR \-m ", " "\-\-queue\-run\-max SEC"
Spend no longer than
.I SEC
seconds processing the change queue before checking
.BR inotify (7)
again, to avoid overflows caused by many changes happening at once.  The
default is 5 seconds.  This will rarely need to be changed.
.TP
.B \-h, \-\-help
Print a usage message on standard output and exit successfully.
.TP
.B \-V, \-\-version 
Print version information on standard output and exit successfully.


.SH NOTES
If you watch a lot of directories, you will probably need to increase the
kernel parameter
.IR fs.inotify.max_user_watches ,
for example:
.PP
.in +4
sysctl -w fs.inotify.max_user_watches=1048576
.in
.PP
You may also need to increase
.IR fs.inotify.max_user_instances .

The change files can be used directly as input to
.BR rsync (1)'s
.I --files-from
option, like this:
.PP
.in +4
rsync --delete -dlptgoDH --files-from=\fIOUTPUTDIR\fR/20140731-112501.1234 \fIDIRECTORY\fR remote:/dir/copy/
.in
.PP
A similar command line is used by
.BR continual-sync (1)
when doing a partial sync.

Changes to file permissions are not listed - only changes which alter the
contents of a file or its last-modification time.


.SH BUGS
When watching a directory with a large number of subdirectories, it may take
.BR watchdir (1)
an inconveniently long time to finish scanning through them all and begin
watching for changes.


.SH REPORTING BUGS
Report bugs in
.B continual-sync
to continual-sync@ivarch.com or use the contact form linked from the package
home page: <http://www.ivarch.com/programs/continual-sync.shtml>


.SH SEE ALSO
.BR continual-sync "(1), " rsync "(1), " glob "(7), " lsyncd "(1), " rsync-inotify "(1), " inosync (1)


.SH LICENSE
This is free software, distributed under the ARTISTIC 2.0 license.


NEWS
0.0.6 - 4 September 2021
  * Added an "ignore vanished files" option
  * Updated the init script to create /var/run/continual-sync on startup

0.0.5 - 12 December 2014
  * repackaged as an Open Source project with permission from my employer

0.0.4 - 12 August 2014
  * added "working directory" field to status file
  * check working directory still exists after every sync attempt, as well as
    when a watcher exits

0.0.3 - 7 August 2014
  * added "current action" field to status file

0.0.2 - 6 August 2014
  * added "temporary directory" parameter
  * set default $PATH if one is not set in the environment
  * record rsync errors in log file
  * added "status file" parameter

0.0.1 - 2 August 2014
  * initial release


watchdir.c
/*
 * Command-line interface to the watch_dir function provided in watch.c.
 */

#include <stdio.h>
#include <limits.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <getopt.h>
#include "common.h"

#define MAX_EXCLUDES 1000

extern int watch_dir(const char *toplevel_path,
		     const char *changedpath_dir,
		     unsigned long full_scan_interval,
		     unsigned long queue_run_interval,
		     unsigned long queue_run_max_seconds,
		     unsigned long changedpath_dump_interval,
		     unsigned int max_dir_depth, char **excludes,
		     unsigned int exclude_count);

/* List of command line parameters after options. */
static char **parameters = NULL;
static int parameter_count = 0;

/* Parameters that can be overridden by command line options. */
static unsigned long full_scan_interval = 7200;
static unsigned long queue_run_interval = 2;
static unsigned long queue_run_max_seconds = 5;
static unsigned long changedpath_dump_interval = 30;
static unsigned int max_dir_depth = 20;
static char *excludes[MAX_EXCLUDES];
static unsigned int exclude_count = 0;


/*
 * Output program usage information.
 */
static void usage(void)
{
	printf("%s: %s %s\n", _("Usage"),
	       common_program_name, _("[OPTIONS] DIRECTORY OUTPUTDIR"));
	printf("%s\n",
	       _
	       ("Watch DIRECTORY for changes, dumping the changed paths to a unique file in\nthe OUTPUTDIR directory every few seconds."));
	printf("\n");
	printf("  -i, --dump-interval %s (%lu)\n",
	       _("SEC       interval between writing change files"),
	       changedpath_dump_interval);
	printf("  -f, --full-scan-interval %s (%lu)\n",
	       _("SEC  do full rescan every SEC seconds"),
	       full_scan_interval);
	printf("  -e, --exclude %s (*.tmp, *~)\n",
	       _("PATTERN         glob pattern to exclude"));
	printf("  -r, --recursion-depth %s (%u)\n",
	       _("NUM     max depth to descend directories"),
	       max_dir_depth);
	printf("  -q, --queue-run-interval %s (%lu)\n",
	       _("SEC  inotify queue processing interval"),
	       queue_run_interval);
	printf("  -m, --queue-run-max %s (%lu)\n",
	       _("SEC       max time to spend processing queue"),
	       queue_run_max_seconds);
	printf("\n");
	printf("  -h, --help     %s\n", _("display this help and exit"));
	printf("  -V, --version  %s\n",
	       _("display program version and exit"));
#if ENABLE_DEBUGGING
	printf("  -d, --debug    %s\n", _("enable debugging"));
#endif
	printf("\n");
	printf("%s\n",
	       _
	       ("The OUTPUTDIR must not be under the DIRECTORY being watched."));
}


/*
 * Parse the command line arguments.  Returns 0 on success, -1 if the
 * program should exit immediately without an error, or 1 if the program
 * should exit with an error.
 */
static int parse_options(int argc, char **argv)
{
	struct option long_options[] = {
		{"help", 0, 0, 'h'},
		{"version", 0, 0, 'V'},
		{"full-scan-interval", 1, 0, 'f'},
		{"full", 1, 0, 'f'},
		{"exclude", 1, 0, 'e'},
		{"recursion-depth", 1, 0, 'r'},
		{"queue-run-interval", 1, 0, 'q'},
		{"queue", 1, 0, 'q'},
		{"queue-run-max", 1, 0, 'm'},
		{"max", 1, 0, 'm'},
		{"dump-interval", 1, 0, 'i'},
		{"interval", 1, 0, 'i'},
		{"depth", 1, 0, 'r'},
#if ENABLE_DEBUGGING
		{"debug", 0, 0, 'd'},
#endif
		{0, 0, 0, 0}
	};
	int option_index = 0;
	char *short_options = "hVf:e:r:q:m:i:"
#if ENABLE_DEBUGGING
	    "d"
#endif
	    ;
	int c;
	unsigned long param;

	parameter_count = 0;
	parameters = calloc(argc + 1, sizeof(char *));
	if (NULL == parameters) {
		die("%s", strerror(errno));
		return 1;
	}

	do {
		c = getopt_long(argc, argv, short_options, long_options,
				&option_index);
		if (c < 0)
			continue;

		/*
		 * Parse each command line option.
		 */
		switch (c) {
		case 'h':
			usage();
			free(parameters);
			parameters = NULL;
			parameter_count = 0;
			return -1;
			break;
		case 'V':
			printf("%s %s\n", common_program_name, VERSION);
			free(parameters);
			parameters = NULL;
			parameter_count = 0;
			return -1;
			break;
#if ENABLE_DEBUGGING
		case 'd':
			debugging_enabled = 1;
			break;
#endif
		case 'e':
			if (exclude_count >= (MAX_EXCLUDES - 1)) {
				error("%s",
				      _
				      ("maximum number of excludes reached"));
				free(parameters);
				parameters = NULL;
				parameter_count = 0;
				return 1;
			}
			excludes[exclude_count++] = xstrdup(optarg);
			break;
		case 'f':
		case 'r':
		case 'q':
		case 'm':
		case 'i':
			errno = 0;
			param = strtoul(optarg, NULL, 10);
			if (0 != errno) {
				error("-%c: %s", c, strerror(errno));
				free(parameters);
				parameters = NULL;
				parameter_count = 0;
				return 1;
			}
			switch (c) {
			case 'f':
				full_scan_interval = param;
				break;
			case 'r':
				max_dir_depth = param;
				break;
			case 'q':
				queue_run_interval = param;
				break;
			case 'm':
				queue_run_max_seconds = param;
				break;
			case 'i':
				changedpath_dump_interval = param;
				break;
			}
			break;
		default:
			fprintf(stderr,
				_("Try `%s --help' for more information."),
				common_program_name);
			fprintf(stderr, "\n");
			free(parameters);
			parameters = NULL;
			parameter_count = 0;
			return 1;
			break;
		}

	} while (c != -1);

	/*
	 * Store remaining command-line arguments.
	 */
	while (optind < argc) {
		parameters[parameter_count++] = argv[optind++];
	}

	if (parameter_count != 2) {
		usage();
		free(parameters);
		parameters = NULL;
		parameter_count = 0;
		return 1;
	}

	return 0;
}


/*
 * Command line entry point: parse command line arguments and start the main
 * watch_dir loop.
 */
int main(int argc, char **argv)
{
	char *toplevel_path;		 /* full path to watched dir */
	char *changedpath_dir;		 /* full path to output queue dir */
	int rc;
	int eidx;

	common_program_name = ds_leafname(argv[0]);

	rc = parse_options(argc, argv);
	if (rc < 0)
		return EXIT_SUCCESS;
	else if (rc > 0)
		return EXIT_FAILURE;

	toplevel_path = realpath(parameters[0], NULL);
	if (NULL == toplevel_path) {
		fprintf(stderr, "%s: %s: %s\n", common_program_name,
			parameters[0], strerror(errno));
		exit(EXIT_FAILURE);
	}

	changedpath_dir = realpath(parameters[1], NULL);
	if (NULL == changedpath_dir) {
		fprintf(stderr, "%s: %s: %s\n", common_program_name,
			parameters[1], strerror(errno));
		free(toplevel_path);
		exit(EXIT_FAILURE);
	}

	rc = watch_dir(toplevel_path, changedpath_dir, full_scan_interval,
		       queue_run_interval, queue_run_max_seconds,
		       changedpath_dump_interval, max_dir_depth, excludes,
		       exclude_count);

	free(toplevel_path);
	free(changedpath_dir);

	for (eidx = 0; eidx < exclude_count; eidx++) {
		if (NULL != excludes[eidx])
			free(excludes[eidx]);
	}
	if (NULL != parameters)
		free(parameters);

	return rc;
}

/* EOF */
